### PR TITLE
DAO Coin Exchange API endpoints

### PR DESCRIPTION
**Pull request by iamsofonias**

### CODE REVIEW

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:22]: We typically use the `EntryResponse` suffix to denote a differentiation between the core struct and the backend struct.

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:23]: do we ever actually return a username here? or is it only ever a base58 check pub key? If it's only ever a pub key, denote it as such. Same applies to the Buying / Selling attributes below.

I think it makes most sense to change these attributes to be Base58Check and include ProfileEntryResponses for each in order to provide username / other details needed.

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:34]: Will a float64 actually hold this? I'm not sure...

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:30]: will a float64 be big enough?

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:87]: can probably refactor this into a function since this is all duplicated below. It can be a function that's defined within this function if that makes some things easier.

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:154]: we need to catch this error. 
```suggestion
	if err = json.NewEncoder(ww).Encode(res); err != nil {
		_AddBadRequestError(ww, fmt.Sprintf("GetDAOCoinLimitOrders: Problem encoding response as JSON: %v", err))
		return
	}
```

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:124]: it might be better to write a function `_daoCoinLimitOrderToResponse` to handle this.

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:169]: probably need to watch out for overflow here. Not sure a float64 is gonna do the job.

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:181]: overflow concerns

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:174]: if the return value of this is only used for frontend consumption, maybe we consider return a string? Worried about precision w/ a float64

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2588]: will a float64 be big enough?

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2689]: lol I think I screwed up the name of this `GetPubKe` function. Can you please fix it as well?

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2709]:
```suggestion
	if requestData.ExchangeRateCoinsToSellPerCoinToBuy != float64(0) {
```

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2634]:
```suggestion
			fmt.Sprint("CreateOrCancelDAOCoinLimitOrder: Must provide exactly one of the "+
```

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2643]:
```suggestion
			fmt.Sprint("CreateOrCancelDAOCoinLimitOrder: Must provide exactly one of the "+
```

**Review by lazynina — CHANGES_REQUESTED** [review]

File: routes/transaction.go
Comment (iamsofonias) [reply routes/transaction.go:2588]: Discussed this offline. Yeah, float64 has a wide enough range

File: routes/transaction.go
Comment (iamsofonias) [reply routes/transaction.go:2709]: Not an issue here. The compiler is smart enough to know it's a float64 type when we use the 0 literal. The code reads a bit cleaner as-is

File: routes/dao_coin_exchange.go
Comment (iamsofonias) [reply routes/dao_coin_exchange.go:180]: Discussed offline. float64 has a wider range than uint256.Int

File: routes/dao_coin_exchange.go
Comment (iamsofonias) [reply routes/dao_coin_exchange.go:192]: Discussed offline. float64 has a wider range than uint256.Int

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:107]: I think we should use `utxoView._getAllDAOCoinLimitOrdersForThisDAOCoinPair` instead of the dbAdapter so we also return results from the mempool

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2694]: should add some error code here before returning - otherwise user will get a 200 response with no constructed tx

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2704]: add some error here before returning

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2728]: remove double if err != nil, only need one

**Review by lazynina — CHANGES_REQUESTED** [review]

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:158]: Can we refactor this into a function called `DAOCoinLimitOrderToResponse`?

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2602]:
```suggestion
	// values and the remaining request params to the helper for further validation
```

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2619]:
```suggestion
	if requestData.QuantityToFill <= 0 {
```

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2686]: See how we parse `postHashToModify` in `SubmitPost` (around line 1307 in transaction.go) for how we parse a block hash hex from the request

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2693]: we may be able to pass nil once the transaction construction function changes and the metadata encoding is fixed.

Comment (lazynina) [reply routes/transaction.go:2693]: or the construction function will be smart and just exclude these values.

File: routes/dao_coin_exchange.go
Comment (lazynina) [inline routes/dao_coin_exchange.go:97]:
```suggestion
	
	getCoinPKIDAndPubKey := func(creatorPubKeyOrUsername string) (*lib.PKID, string, error) {
		coinPKID := &lib.ZeroPKID
		coinPublicKeyBase58Check := ""
		if creatorPubKeyOrUsername != "" {
			coinPublicKeyBase58Check, coinPKID, err = fes.validateCreatorPublicKeyBase58CheckOrUsername(
				utxoView,
				requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername,
			)
			if err != nil {
				return nil, "", nil
			}
		}
		return coinPKID, coinPublicKeyBase58Check, nil
	}
	
	coin1PKID, coin1ProfilePublicBase58Check, err := getCoinPKIDAndPubKey(requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername)
	if err != nil {
		_AddBadRequestError(
			ww,
			fmt.Sprintf("GetDAOCoinLimitOrders: Invalid DAOCoin1CreatorPublicKeyBase58CheckOrUsername: %v", err),
		)
	}

	coin2PKID, coin2ProfilePublicBase58Check, err := getCoinPKIDAndPubKey(requestData.DAOCoin1CreatorPublicKeyBase58CheckOrUsername)
	if err != nil {
		_AddBadRequestError(
			ww,
			fmt.Sprintf("GetDAOCoinLimitOrders: Invalid DAOCoin2CreatorPublicKeyBase58CheckOrUsername: %v", err),
		)
	}
```

Comment (lazynina) [reply routes/dao_coin_exchange.go:97]: refactor into a locally scoped function

**Review by lazynina — CHANGES_REQUESTED** [review]

File: routes/transaction.go
Comment (iamsofonias) [reply routes/transaction.go:2693]: Yeah, these can be nill once MF's PR is merged

File: routes/dao_coin_exchange.go
Comment (iamsofonias) [reply routes/dao_coin_exchange.go:97]: Tried making this change locally. But TBH, I think the code reads more clearly as-is given a locally scoped function adds another level of indirection & nesting without deduplicating code any further. If you don't have strong feelings about this, my preference is to leave it as is.

Comment (lazynina) [reply routes/dao_coin_exchange.go:97]: Fine by me - don't feel strongly about this

File: routes/utils.go
Comment (iamsofonias) [inline routes/utils.go:11]: Moving forward, we should be using this function everywhere we decode block hashes from hex strings. I'll have another PR that migrates existing uses to use this function.

**Comment by iamsofonias** [comment]
This is using Matt Foley's core branch https://github.com/deso-protocol/core/pull/329/files to test the endpoints against latest transaction metadata behavior with optional fields. 

The E2E testing looks good, but I'll hold off on merging until the core PR is merged into main too.

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2760]: should we initialize these to empty byte slices instead so we can exclude them from txn construction? Lib.NewPublicKey returns nil if the number of bytes isn't 33
```suggestion
	var buyingCoinPublicKey []byte
	var sellingCoinPublicKey []byte
```

File: routes/transaction.go
Comment (lazynina) [inline routes/transaction.go:2791]:
```suggestion
```
it's initialized to nil so we can delete this line.

**Review by lazynina — CHANGES_REQUESTED** [review]

File: routes/transaction.go
Comment (iamsofonias) [reply routes/transaction.go:2760]: Ah, that's a good point. We'll want the public keys to be nil for order cancels. But we'll also want them to default to ZeroPublicKey when creating an order that's buying or selling DESO. I'll do a small refactor to account for this.

**Review by lazynina — APPROVED** [review]

**Pull request merged**
